File: src/ssh/keygen.rs
use std::path::{Path, PathBuf};
use std::process::Command;

use crate::providers::KeyType;

#[allow(dead_code)]
pub struct KeyPair {
    pub private_key: PathBuf,
    pub public_key: PathBuf,
}

pub fn generate_key(
    ssh_dir: &Path,
    key_name: &str,
    email: &str,
    key_type: KeyType,
) -> Result<KeyPair, String> {
    let private_key = ssh_dir.join(key_name);
    let public_key = ssh_dir.join(format!("{key_name}.pub"));

    let mut args = vec![
        "-t".to_string(),
        key_type.algorithm().to_string(),
        "-C".to_string(),
        email.to_string(),
        "-f".to_string(),
        private_key.to_string_lossy().to_string(),
        "-N".to_string(),
        String::new(), // empty passphrase
    ];

    if let Some(bits) = key_type.bits() {
        args.push("-b".to_string());
        args.push(bits.to_string());
    }

    let output = Command::new("ssh-keygen")
        .args(&args)
        .output()
        .map_err(|e| format!("Failed to run ssh-keygen: {e}"))?;

    if !output.status.success() {
        // Cleanup partial files on failure
        let _ = std::fs::remove_file(&private_key);
        let _ = std::fs::remove_file(&public_key);
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("ssh-keygen failed: {stderr}"));
    }

    Ok(KeyPair {
        private_key,
        public_key,
    })
}

pub fn key_exists(ssh_dir: &Path, key_name: &str) -> bool {
    ssh_dir.join(key_name).exists()
}

pub fn read_public_key(ssh_dir: &Path, key_name: &str) -> Result<String, String> {
    let pub_path = ssh_dir.join(format!("{key_name}.pub"));
    std::fs::read_to_string(&pub_path)
        .map(|s| s.trim().to_string())
        .map_err(|e| format!("Failed to read {}: {e}", pub_path.display()))
}

pub fn remove_key_files(ssh_dir: &Path, key_name: &str) -> Result<(), String> {
    let private_key = ssh_dir.join(key_name);
    let public_key = ssh_dir.join(format!("{key_name}.pub"));

    let mut errors = Vec::new();

    if private_key.exists() {
        if let Err(e) = std::fs::remove_file(&private_key) {
            errors.push(format!("Failed to remove {}: {e}", private_key.display()));
        }
    }

    if public_key.exists() {
        if let Err(e) = std::fs::remove_file(&public_key) {
            errors.push(format!("Failed to remove {}: {e}", public_key.display()));
        }
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(errors.join("; "))
    }
}

File: src/ssh/keygen.ts
import { existsSync, readFileSync, unlinkSync } from "node:fs";
import { join } from "node:path";
import { spawnSync } from "node:child_process";
import { type ProviderKey, PROVIDERS } from "../providers";
import { commandExists } from "../platform";
import { SSH_DIR } from "./config";

export function generateKey(provider: ProviderKey, email: string, keyName: string): void {
  const keyPath = join(SSH_DIR, keyName);

  if (!commandExists("ssh-keygen")) {
    throw new Error("ssh-keygen is required but not found on PATH.");
  }

  const args =
    PROVIDERS[provider].keyType === "rsa"
      ? ["-t", "rsa", "-b", "4096", "-C", email, "-f", keyPath, "-N", ""]
      : ["-t", "ed25519", "-C", email, "-f", keyPath, "-N", ""];

  const result = spawnSync("ssh-keygen", args, { stdio: "inherit", encoding: "utf8" });
  if (result.error) {
    throw result.error;
  }

  if ((result.status ?? 1) !== 0) {
    throw new Error("ssh-keygen failed.");
  }
}

export function readPublicKey(keyName: string): string {
  const publicKeyPath = join(SSH_DIR, `${keyName}.pub`);
  if (!existsSync(publicKeyPath)) {
    throw new Error(`Public key not found: ${publicKeyPath}`);
  }

  return readFileSync(publicKeyPath, "utf8").trim();
}

export function keyExists(keyName: string): boolean {
  return existsSync(join(SSH_DIR, keyName));
}

export function deleteKeyFiles(keyName: string): void {
  const keyPath = join(SSH_DIR, keyName);

  try {
    unlinkSync(keyPath);
  } catch {
    // Ignore missing private key.
  }

  try {
    unlinkSync(`${keyPath}.pub`);
  } catch {
    // Ignore missing public key.
  }
}
